### DISCUSSION TITLE
OpenSSL 3 reported to cause problems in later Linux Distros using LD_PRELOAD for httpd acceptance of expired client certificates

### DISCUSSION THREAD

**Discussion by jmcpheters** [openssl/openssl#26218] [Q&A] [answered]
My organization has a file server application with TLS requiring client certificates from devices out in the field. The devices have survived longer than expected, and now their client certs are expiring but we cannot replace them and we still need the clients to connect using TLS. The file server uses httpd (or apache2) on a Linux distro. To accept expired client certficates we use LD_PRELOAD to change the default client verify behavior to treat X509_V_ERR_CERT_HAS_EXPIRED as X509_V_OK using the method described at this link:

https://blog.fem.tu-ilmenau.de/archives/1025-Accepting-expired-client-certificates-with-apache-2.4.html

This has worked well for us, until upgrading to the latest Linux distros, and I’ve been told that the reason it stopped working is because of OpenSSL 3. Here is a summary of were the method works and where it stops being functional:

| LD_PRELOAD works? | OpenSSL ver | Linux Distro      | httpd ver |
| ------------------| ------------|-------------------|-----------|
| yes               | 1.1.1       | Amazon Linux 1    | 2.2       |              
| yes               | 1.1.1       | Amazon Linux 2    | 2.4       |
| yes               | 1.1.1       | Ubuntu 18.04      | 2.4       |
| no                | 3.0         | Amazon Linux 2023 | 2.4       |
| no                | 3.0         | Ubuntu 24.04      | 2.4       |

For both Amazon Linux 2023 and Ubuntu 24.04, the LD_PRELOAD method works intermittently, which I estimate to be 40% of the time. So if the file server reboots or httpd restarts, when httpd is running 40% of the time it will come up such that expired certificates are accepted, and those expired certificates will be accepted until the next reboot or restart. Obviously an intermittently working configuration is not manageable.

I’ve previously posted to the Apache forum asking about expired client cert acceptance in later Linux distros, but received no input or answers. At this point we are looking at an SSL proxy that can be configured to allow expired certificates, but we’d like to keep things as simple as possible and just stick with the LD_PRELOAD method if possible.

Does anyone know why OpenSSL 3.0 in the later Linux distros will cause intermittent issues with LD_PRELOAD precluding httpd acceptance expired client certificates? Does anyone know configure OpenSSL 3.0 so that LD_PRELOAD will work again to get httpd to accept expired client certificates?

**Comment by nhorman** [comment, marked as answer]
For the record, doing this seems like a really bad idea.  If you're certificate is expired, the correct thing to do is update your certificate.  But you know that, and if you're aware of the problem, there should be no reason that LD_PRELOAD won't do  what you want it to here.

If I had to guess, the problem would be in whoever registers the "real" verification callback.  If that has updated (or if theres some timing variance that affects which call back gets registered) that could cause your problem (note that in the example you provided, you set orig_verify_callback to whoever calls SSL_CTX_set_verify first, even though this callback is a per-CTX pointer).  So if you have multiple CTX objects each with a unique verify_callback, this is going to funnel all of them into the callback for whoever registered "first").  I'd start by adding some debug printfs to the preloaded object to see what the CTX callback is vs which one you're calling

**Reply by jmcpheters** [reply]
This:

> So if you have multiple CTX objects each with a unique verify_callback , this is going to funnel all of them into the callback for whoever registered "first"

I changed the code to tie the verify callback to the SSL_CTX. With this change, the LD_PRELOAD with the updated .so now works 100% of the time.

**Comment by mattcaswell** [comment]
My guess would be some race condition is happening. I can't think of anything in particular that would cause this with 3.0. I would instrument the preloaded coded to see if it actually gets called at all in the problematic cases.
